From bfaec70aecde28f8232ac182cea7ed9537a5d75f Mon Sep 17 00:00:00 2001
From: Alexander Hoffman <alxhoff@gmail.com>
Date: Fri, 28 Nov 2025 14:17:08 +0100
Subject: [PATCH 05/20] D4xx i2c sleeps

---
 kernel/nvidia/drivers/media/i2c/d4xx.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/kernel/nvidia/drivers/media/i2c/d4xx.c b/kernel/nvidia/drivers/media/i2c/d4xx.c
index 7fabaf382..04c8969ec 100644
--- a/kernel/nvidia/drivers/media/i2c/d4xx.c
+++ b/kernel/nvidia/drivers/media/i2c/d4xx.c
@@ -516,6 +516,8 @@ static int ds5_write_8(struct ds5 *state, u16 reg, u8 val)
 			dev_dbg(&state->client->dev, "%s(): i2c write 0x%04x: 0x%x\n",
 				 __func__, reg, val);
 
+	usleep_range(100,110);
+
 	return ret;
 }
 #endif
@@ -542,6 +544,8 @@ static int ds5_write(struct ds5 *state, u16 reg, u16 val)
 			dev_dbg(&state->client->dev, "%s(): i2c write 0x%04x: 0x%x\n",
 				__func__, reg, val);
 
+	usleep_range(100,110);
+
 	return ret;
 }
 
@@ -558,7 +562,7 @@ static int ds5_raw_write(struct ds5 *state, u16 reg,
 			dev_dbg(&state->client->dev,
 					"%s(): i2c raw write 0x%04x: %d bytes\n",
 					__func__, reg, (int)val_len);
-
+	usleep_range(100,110);
 	return ret;
 }
 
@@ -573,7 +577,7 @@ static int ds5_read(struct ds5 *state, u16 reg, u16 *val)
 			dev_dbg(&state->client->dev, "%s(): i2c read 0x%04x: 0x%x\n",
 					__func__, reg, *val);
 	}
-
+	usleep_range(100,110);
 	return ret;
 }
 
@@ -584,6 +588,8 @@ static int ds5_raw_read(struct ds5 *state, u16 reg, void *val, size_t val_len)
 		dev_err(&state->client->dev, "%s(): i2c read failed %d, 0x%04x\n",
 			__func__, ret, reg);
 
+	usleep_range(100,110);
+
 	return ret;
 }
 
-- 
2.51.0

