From 082ae68facb53cbc32cba0b340d071322d12cb9d Mon Sep 17 00:00:00 2001
From: Alexander Hoffman <alxhoff@gmail.com>
Date: Mon, 7 Apr 2025 14:38:05 +0200
Subject: [PATCH 14/33] Printing stack trace on ds5_read/write failure

---
 kernel/nvidia/drivers/media/i2c/d4xx.c | 15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/kernel/nvidia/drivers/media/i2c/d4xx.c b/kernel/nvidia/drivers/media/i2c/d4xx.c
index 04c8969ec..2b10ced0c 100644
--- a/kernel/nvidia/drivers/media/i2c/d4xx.c
+++ b/kernel/nvidia/drivers/media/i2c/d4xx.c
@@ -30,6 +30,7 @@
 #include <linux/string.h>
 #include <linux/videodev2.h>
 #include <linux/version.h>
+#include <linux/stacktrace.h>
 #ifdef CONFIG_VIDEO_INTEL_IPU6
 #include <linux/ipu-isys.h>
 #include <media/d4xx_pdata.h>
@@ -535,14 +536,19 @@ static int ds5_write(struct ds5 *state, u16 reg, u16 val)
 			__func__, reg, value[1], value[0]);
 
 	ret = regmap_raw_write(state->regmap, reg, value, sizeof(value));
-	if (ret < 0)
+	if (ret < 0) {
+		dev_err(&state->client->dev,
+				"%s(): writing to register: 0x%04x, value1: 0x%x, value2:0x%x\n",
+				__func__, reg, value[1], value[0]);
 		dev_err(&state->client->dev,
 				"%s(): i2c write failed %d, 0x%04x = 0x%x\n",
 				__func__, ret, reg, val);
-	else
+		stack_trace_print(NULL, 0, 4);
+	} else {
 		if (state->dfu_dev.dfu_state_flag == DS5_DFU_IDLE)
 			dev_dbg(&state->client->dev, "%s(): i2c write 0x%04x: 0x%x\n",
 				__func__, reg, val);
+	}
 
 	usleep_range(100,110);
 
@@ -569,10 +575,11 @@ static int ds5_raw_write(struct ds5 *state, u16 reg,
 static int ds5_read(struct ds5 *state, u16 reg, u16 *val)
 {
 	int ret = regmap_raw_read(state->regmap, reg, val, 2);
-	if (ret < 0)
+	if (ret < 0){
 		dev_err(&state->client->dev, "%s(): i2c read failed %d, 0x%04x\n",
 				__func__, ret, reg);
-	else {
+		stack_trace_print(NULL, 0, 4);
+	} else {
 		if (state->dfu_dev.dfu_state_flag == DS5_DFU_IDLE)
 			dev_dbg(&state->client->dev, "%s(): i2c read 0x%04x: 0x%x\n",
 					__func__, reg, *val);
-- 
2.52.0

