diff --git a/kernel/nvidia/drivers/media/i2c/d4xx.c b/kernel/nvidia/drivers/media/i2c/d4xx.c
index 04c8969ec..e5874f493 100644
--- a/kernel/nvidia/drivers/media/i2c/d4xx.c
+++ b/kernel/nvidia/drivers/media/i2c/d4xx.c
@@ -30,6 +30,7 @@
 #include <linux/string.h>
 #include <linux/videodev2.h>
 #include <linux/version.h>
+#include <linux/stacktrace.h>
 #ifdef CONFIG_VIDEO_INTEL_IPU6
 #include <linux/ipu-isys.h>
 #include <media/d4xx_pdata.h>
@@ -562,22 +563,27 @@ static int ds5_raw_write(struct ds5 *state, u16 reg,
 			dev_dbg(&state->client->dev,
 					"%s(): i2c raw write 0x%04x: %d bytes\n",
 					__func__, reg, (int)val_len);
 	usleep_range(100,110);
 	return ret;
 }
 
 static int ds5_read(struct ds5 *state, u16 reg, u16 *val)
 {
 	int ret = regmap_raw_read(state->regmap, reg, val, 2);
-	if (ret < 0)
+	if (ret < 0){
 		dev_err(&state->client->dev, "%s(): i2c read failed %d, 0x%04x\n",
 				__func__, ret, reg);
-	else {
+		stack_trace_print(NULL, 0, 4);
+	} else {
 		if (state->dfu_dev.dfu_state_flag == DS5_DFU_IDLE)
 			dev_dbg(&state->client->dev, "%s(): i2c read 0x%04x: 0x%x\n",
 					__func__, reg, *val);
 	}
 	usleep_range(100,110);
 	return ret;
 }
 
