diff --git a/kernel/nvidia/drivers/media/i2c/max9295.c b/kernel/nvidia/drivers/media/i2c/max9295.c
index 6311d97d7..f354739e2 100644
--- a/kernel/nvidia/drivers/media/i2c/max9295.c
+++ b/kernel/nvidia/drivers/media/i2c/max9295.c
@@ -573,6 +573,27 @@ int max9295_init_settings(struct device *dev)
 		{MAX9295_START_PIPE_ADDR, 0xF0},
 	};
 
+	static struct reg_pair map_depth_trigger[] = {
+	//	{0x02D6, 0x84}, // #MFP8
+	//	{0x02D7, 0x60}, // #OUT_TYPE bit to 1
+	//	{0x02D8, 0x1F},
+		{0x02C1, 0x84}, // #MFP1
+		{0x02C2, 0x20}, // #OUT_TYPE bit to 1 (no pullup)
+		{0x02C3, 0x1F},
+	};
+
+	static struct reg_pair map_rgb_trigger[] = {
+		{0x02BE, 0x84}, // #MFP0
+		{0x02BF, 0x20}, // #OUT_TYPE bit to 1  (no pullup)
+		{0x02C0, 0x1B},
+	};
+
+	if (!priv) {
+		goto ret;
+	}
+
+    dev_info(dev, "%s starting init\n", __func__);
+
 	mutex_lock(&priv->lock);
 
 	// Init control
@@ -583,6 +604,13 @@ int max9295_init_settings(struct device *dev)
 		err |= __max9295_set_pipe(dev, i, GMSL_CSI_DT_YUV422_8,
 					  GMSL_CSI_DT_EMBED, i);
 
+	// Trigger Depth
+	err |= max9295_set_registers(dev, map_depth_trigger,
+				     ARRAY_SIZE(map_depth_trigger));
+	// Trigger RGB
+	err |= max9295_set_registers(dev, map_rgb_trigger,
+				     ARRAY_SIZE(map_rgb_trigger));
+
 	mutex_unlock(&priv->lock);
 
 	return err;
diff --git a/kernel/nvidia/drivers/media/i2c/max9296.c b/kernel/nvidia/drivers/media/i2c/max9296.c
index 316e687a1..7cf01e063 100644
--- a/kernel/nvidia/drivers/media/i2c/max9296.c
+++ b/kernel/nvidia/drivers/media/i2c/max9296.c
@@ -932,12 +932,34 @@ int max9296_init_settings(struct device *dev)
 	int i;
 	struct max9296 *priv = dev_get_drvdata(dev);
 
+	static struct reg_pair map_depth_trigger[] = {
+		// For GMSL_10CH_first board (isx031 and d457 use one trig gpio)
+		{0x02C2, 0x82}, // #MFP6
+		{0x02C3, 0x1F},
+		// For GMSL_10CH_SCH_V1.1 version (isx031 and d457 use diffrent trig gpio )
+		// {0x02C5, 0x82}, #MFP7
+		// {0x02C6, 0x1F},
+	};
+
+	static struct reg_pair map_rgb_trigger[] = {
+		{0x02CB, 0x82}, // #MFP9
+		{0x02CC, 0x1B},
+	};
+
 	mutex_lock(&priv->lock);
 
 	for (i = 0; i < MAX9296_MAX_PIPES; i++)
 		err |= __max9296_set_pipe(dev, i, GMSL_CSI_DT_YUV422_8,
 					  GMSL_CSI_DT_EMBED, i);
 
+	// Trigger Depth
+	err |= max9296_set_registers(dev, map_depth_trigger,
+				    ARRAY_SIZE(map_depth_trigger));
+	// Trigger RGB
+	err |= max9296_set_registers(dev, map_rgb_trigger,
+                    ARRAY_SIZE(map_rgb_trigger));
+	
+
 	mutex_unlock(&priv->lock);
 
 	return err;
