diff --git a/drivers/media/platform/tegra/camera/vi/vi5_fops.c b/drivers/media/platform/tegra/camera/vi/vi5_fops.c
index 80b7d2bd9540745b08e5f0d483c33af3be6e7a07..2071b368732cb0348fb29636ab1be1adaf88bd56 100644 (file)
--- a/drivers/media/platform/tegra/camera/vi/vi5_fops.c
+++ b/drivers/media/platform/tegra/camera/vi/vi5_fops.c
@@ -485,6 +485,7 @@ static void vi5_capture_dequeue(struct tegra_channel *chan,
        struct tegra_channel_buffer *buf)
 {
        int err = 0;
+       bool frame_err = false;
        int vi_port = 0;
        int gang_prev_frame_id = 0;
        unsigned long flags;
@@ -528,12 +529,7 @@ static void vi5_capture_dequeue(struct tegra_channel *chan,
                                        "err_data %d\n",
                                        descr->status.frame_id, descr->status.flags,
                                        descr->status.err_data);
-#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 4, 0)
-                                       buf->vb2_state = VB2_BUF_STATE_REQUEUEING;
-#else
-                                       buf->vb2_state = VB2_BUF_STATE_ERROR;
-#endif
-                       goto done;
+                               frame_err = true;
                        }
                } else if (!vi_port) {
                        gang_prev_frame_id = descr->status.frame_id;
@@ -565,7 +561,14 @@ static void vi5_capture_dequeue(struct tegra_channel *chan,
 #endif
        vb->vb2_buf.timestamp = descr->status.sof_timestamp;
 
-       buf->vb2_state = VB2_BUF_STATE_DONE;
+       if (frame_err)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 4, 0)
+               buf->vb2_state = VB2_BUF_STATE_REQUEUEING;
+#else
+               buf->vb2_state = VB2_BUF_STATE_ERROR;
+#endif
+       else
+               buf->vb2_state = VB2_BUF_STATE_DONE;
        /* Read EOF from capture descriptor */
 #if LINUX_VERSION_CODE < KERNEL_VERSION(5, 4, 0)
        ts = ns_to_timespec((s64)descr->status.eof_timestamp);
@@ -578,7 +581,6 @@ static void vi5_capture_dequeue(struct tegra_channel *chan,
        trace_tegra_channel_capture_frame("eof", &ts);
 #endif
 
-done:
        goto rel_buf;
 
 uncorr_err:
