diff --git a/hardware/nvidia/platform/t23x/concord/kernel-dts/cvb/tegra234-p3737-0000-camera-d4xx-quad-sw-v1_1.dtsi b/hardware/nvidia/platform/t23x/concord/kernel-dts/cvb/tegra234-p3737-0000-camera-d4xx-quad-sw-v1_1.dtsi
index 5375712f0..d959644ba 100644
--- a/hardware/nvidia/platform/t23x/concord/kernel-dts/cvb/tegra234-p3737-0000-camera-d4xx-quad-sw-v1_1.dtsi
+++ b/hardware/nvidia/platform/t23x/concord/kernel-dts/cvb/tegra234-p3737-0000-camera-d4xx-quad-sw-v1_1.dtsi
@@ -37,7 +37,7 @@
 	};
 
 	tegra-capture-vi {
-		num-channels = <14>;
+		num-channels = <16>;
 
 		ports {
 			#address-cells = <0x1>;
@@ -47,48 +47,48 @@
 				reg = <0>;
 				status = "ok";
 
-				vi_in0: endpoint {
+				d4m0_depth_vi_in: endpoint {
 					status = "ok";
 					port-index = <4>;
 					bus-width = <2>;
 					vc-id = <0>;
-					remote-endpoint = <&csi_out0>;
+					remote-endpoint = <&d4m0_depth_csi_out>;
 				};
 			};
 			port@1 {
 				reg = <1>;
 				status = "ok";
 
-				vi_in1: endpoint {
+				d4m0_ir_vi_in: endpoint {
 					status = "ok";
 					port-index = <4>;
 					bus-width = <2>;
 					vc-id = <1>;
-					remote-endpoint = <&csi_out1>;
+					remote-endpoint = <&d4m0_ir_csi_out>;
 				};
 			};
 			port@2 {
 				reg = <2>;
 				status = "ok";
 
-				vi_in2: endpoint {
+				d4m1_depth_vi_in: endpoint {
 					status = "ok";
 					port-index = <4>;
 					bus-width = <2>;
 					vc-id = <2>;
-					remote-endpoint = <&csi_out2>;
+					remote-endpoint = <&d4m1_depth_csi_out>;
 				};
 			};
 			port@3 {
 				reg = <3>;
 				status = "ok";
 
-				vi_in3: endpoint {
+				d4m1_ir_vi_in: endpoint {
 					status = "ok";
 					port-index = <4>;
 					bus-width = <2>;
 					vc-id = <3>;
-					remote-endpoint = <&csi_out3>;
+					remote-endpoint = <&d4m1_ir_csi_out>;
 				};
 			};
 			//SG10A_ORIN_GMSL2 adapt board port J23/J24 for D457
@@ -96,103 +96,134 @@
 				reg = <4>;
 				status = "ok";
 
-				vi_in4: endpoint {
+				d4m2_depth_vi_in: endpoint {
 					status = "ok";
 					port-index = <2>;
 					bus-width = <2>;
 					vc-id = <0>;
-					remote-endpoint = <&csi_out4>;
+					remote-endpoint = <&d4m2_depth_csi_out>;
 				};
 			};
 			port@5 {
 				reg = <5>;
 				status = "ok";
 
-				vi_in5: endpoint {
+				d4m2_ir_vi_in: endpoint {
 					status = "ok";
 					port-index = <2>;
 					bus-width = <2>;
 					vc-id = <1>;
-					remote-endpoint = <&csi_out5>;
+					remote-endpoint = <&d4m2_ir_csi_out>;
 				};
 			};
 			port@6 {
 				reg = <6>;
 				status = "ok";
 
-				vi_in6: endpoint {
+				d4m3_depth_vi_in: endpoint {
 					status = "ok";
 					port-index = <2>;
 					bus-width = <2>;
 					vc-id = <2>;
-					remote-endpoint = <&csi_out6>;
+					remote-endpoint = <&d4m3_depth_csi_out>;
 				};
 			};
 			port@7 {
 				reg = <7>;
 				status = "ok";
 
-				vi_in7: endpoint {
+				d4m3_ir_vi_in: endpoint {
 					status = "ok";
 					port-index = <2>;
 					bus-width = <2>;
 					vc-id = <3>;
-					remote-endpoint = <&csi_out7>;
+					remote-endpoint = <&d4m3_ir_csi_out>;
 				};
 			};
-			//isx031 camera
+			// Realsense #4&5
 			port@8 {
 				reg = <8>;
-				fzcam_vi_in0: endpoint {
+				status = "ok";
+
+				d4m4_depth_vi_in: endpoint {
+					status = "ok";
+					port-index = <3>;
+					bus-width = <2>;
 					vc-id = <0>;
-					port-index = <0>;
-					bus-width = <4>;
-					remote-endpoint = <&fzcam_csi_out0>;
+					remote-endpoint = <&d4m4_depth_csi_out>;
 				};
 			};
 			port@9 {
 				reg = <9>;
-				fzcam_vi_in1: endpoint {
+				status = "ok";
+
+				d4m4_ir_vi_in: endpoint {
+					status = "ok";
+					port-index = <3>;
+					bus-width = <2>;
 					vc-id = <1>;
-					port-index = <0>;
-					bus-width = <4>;
-					remote-endpoint = <&fzcam_csi_out1>;
+					remote-endpoint = <&d4m4_ir_csi_out>;
 				};
 			};
 			port@10 {
 				reg = <10>;
-				fzcam_vi_in2: endpoint {
+				status = "ok";
+
+				d4m5_depth_vi_in: endpoint {
+					status = "ok";
+					port-index = <3>;
+					bus-width = <2>;
+					vc-id = <2>;
+					remote-endpoint = <&d4m5_depth_csi_out>;
+				};
+			};
+			port@11 {
+				reg = <11>;
+				status = "ok";
+
+				d4m5_ir_vi_in: endpoint {
+					status = "ok";
+					port-index = <3>;
+					bus-width = <2>;
+					vc-id = <3>;
+					remote-endpoint = <&d4m5_ir_csi_out>;
+				};
+			};
+			//isx031 camera
+			port@12 {
+				reg = <12>;
+				isx031_0_vi_in: endpoint {
 					vc-id = <0>;
 					port-index = <5>;
 					bus-width = <4>;
-					remote-endpoint = <&fzcam_csi_out2>;
+					remote-endpoint = <&isx031_0_csi_out>;
 				};
 			};
-			port@11 {
-				reg = <11>;
-				fzcam_vi_in3: endpoint {
+			port@13 {
+				reg = <13>;
+				isx031_1_vi_in: endpoint {
 					vc-id = <1>;
 					port-index = <5>;
 					bus-width = <4>;
-					remote-endpoint = <&fzcam_csi_out3>;
+					remote-endpoint = <&isx031_1_csi_out>;
 				};
 			};
-			port@12 {
-				reg = <12>;
-				fzcam_vi_in4: endpoint {
+			port@14 {
+				reg = <14>;
+				isx031_2_vi_in: endpoint {
 					vc-id = <2>;
 					port-index = <5>;
 					bus-width = <4>;
-					remote-endpoint = <&fzcam_csi_out4>;
+					remote-endpoint = <&isx031_2_csi_out>;
 				};
 			};
-			port@13 {
-				reg = <13>;
-				fzcam_vi_in5: endpoint {
+			port@15 {
+				reg = <15>;
+				isx031_3_vi_in: endpoint {
 					vc-id = <3>;
 					port-index = <5>;
 					bus-width = <4>;
-					remote-endpoint = <&fzcam_csi_out5>;
+					remote-endpoint = <&isx031_3_csi_out>;
 				};
 			};
 		};
@@ -214,7 +245,7 @@
 			force_bus_start = <CAMERA_I2C_MUX_BUS(0)>;
 			vcc_lp = "vcc";
 
-			i2c@0 {
+			i2c@0 { //MAX9296 #0 - 2xD4XX
 				reg = <0x0>;
 				#address-cells = <0x1>;
 				#size-cells = <0x0>;
@@ -267,10 +298,10 @@
 
 						port@0 {
 							reg = <0>;
-							d4m0_out: endpoint {
+							d4m0_depth_out: endpoint {
 								port-index = <4>;
 								bus-width = <2>;
-								remote-endpoint = <&csi_in0>;
+								remote-endpoint = <&d4m0_depth_csi_in>;
 							};
 						};
 					};
@@ -314,10 +345,10 @@
 
 						port@0 {
 							reg = <0>;
-							d4m1_out: endpoint {
+							d4m0_ir_out: endpoint {
 								port-index = <4>;
 								bus-width = <2>;
-								remote-endpoint = <&csi_in1>;
+								remote-endpoint = <&d4m0_ir_csi_in>;
 							};
 						};
 					};
@@ -362,10 +393,10 @@
 
 						port@0 {
 							reg = <0>;
-							d4m2_out: endpoint {
+							d4m1_depth_out: endpoint {
 								port-index = <4>;
 								bus-width = <2>;
-								remote-endpoint = <&csi_in2>;
+								remote-endpoint = <&d4m1_depth_csi_in>;
 							};
 						};
 					};
@@ -409,10 +440,10 @@
 
 						port@0 {
 							reg = <0>;
-							d4m3_out: endpoint {
+							d4m1_ir_out: endpoint {
 								port-index = <4>;
 								bus-width = <2>;
-								remote-endpoint = <&csi_in3>;
+								remote-endpoint = <&d4m1_ir_csi_in>;
 							};
 						};
 					};
@@ -441,7 +472,7 @@
 				};
 			};
 
-			i2c@1 {
+			i2c@1 { //MAX9296 #1 - 2xD4XX
 				reg = <0x1>;
 				#address-cells = <0x1>;
 				#size-cells = <0x0>;
@@ -493,10 +524,10 @@
 						#size-cells = <0>;
 						port@0 {
 							reg = <0>;
-							d4m4_out: endpoint {
+							d4m2_depth_out: endpoint {
 								port-index = <2>;
 								bus-width = <2>;
-								remote-endpoint = <&csi_in4>;
+								remote-endpoint = <&d4m2_depth_csi_in>;
 							};
 						};
 					};
@@ -540,10 +571,10 @@
 
 						port@0 {
 							reg = <0>;
-							d4m5_out: endpoint {
+							d4m2_ir_out: endpoint {
 								port-index = <2>;
 								bus-width = <2>;
-								remote-endpoint = <&csi_in5>;
+								remote-endpoint = <&d4m2_ir_csi_in>;
 							};
 						};
 					};
@@ -588,10 +619,10 @@
 
 						port@0 {
 							reg = <0>;
-							d4m6_out: endpoint {
+							d4m3_depth_out: endpoint {
 								port-index = <2>;
 								bus-width = <2>;
-								remote-endpoint = <&csi_in6>;
+								remote-endpoint = <&d4m3_depth_csi_in>;
 							};
 						};
 					};
@@ -635,10 +666,10 @@
 
 						port@0 {
 							reg = <0>;
-							d4m7_out: endpoint {
+							d4m3_ir_out: endpoint {
 								port-index = <2>;
 								bus-width = <2>;
-								remote-endpoint = <&csi_in7>;
+								remote-endpoint = <&d4m3_ir_csi_in>;
 							};
 						};
 					};
@@ -667,162 +698,233 @@
 				};
 			};
 
-			i2c@2 {
-				reg = <2>;
-				#address-cells = <1>;
-				#size-cells = <0>;
+			i2c@2 { //MAX9296 #2 - 2xD4XX reserve
+				reg = <0x2>;
+				#address-cells = <0x1>;
+				#size-cells = <0x0>;
 				i2c-mux,deselect-on-exit;
 				
-				serdes@48 {
-					compatible = "nvidia,serdes";
+				dser2: max9296@48 {
+					status = "ok";
 					reg = <0x48>;
-					pwdn-gpios = <&tegra_main_gpio CAM2_PWDN GPIO_ACTIVE_HIGH>;
-					channel = <0>;
+					compatible = "maxim,max9296";
+					csi-mode = "2x4";
+					max-src = <2>;
+					reset-gpios = <&tegra_main_gpio CAM2_PWDN GPIO_ACTIVE_HIGH>;
 				};
 
-				fzcam_a@2a {
-					compatible = "nvidia,fzcam";
-					reg = <0x2a>;
+				ser2_prim: max9295_prim@40 {
+					status = "ok";
+					reg = <0x40>;
+					compatible = "maxim,max9295";
+                    is-prim-ser;
+				};
 
-					devnode = "video0";
-					sensor_model ="fzcam";
+				ser2_a: max9295_a@42 {
+					compatible = "maxim,max9295";
+                    reg = <0x42>;
+					maxim,gmsl-dser-device = <&dser2>;
+                    prim-ser = <&ser2_prim>;
+				};
 
-					clocks = <&bpmp_clks TEGRA234_CLK_EXTPERIPH1>,
-							<&bpmp_clks TEGRA234_CLK_EXTPERIPH1>;
-					clock-names = "extperiph1", "pllp_grtba";
-					mclk = "extperiph1";
-					channel = <0>;
-					
+				ser2_b: max9295_b@60 {
+					compatible = "maxim,max9295";
+					reg = <0x60>;
+					maxim,gmsl-dser-device = <&dser2>;
+                    prim-ser = <&ser2_prim>;
+				};
+
+				d4m4_depth: d4m4_depth@0a {
+					status = "ok";
+					def-addr = <0x10>;
+					def-d4m-addr = <0x1c>;
+					reg = <0x0a>;
+					compatible = "intel,d4xx";
+					vcc-supply = <&p3737_vdd_1v8_sys>;
+					cam-type = "Depth";
+					maxim,gmsl-ser-device = <&ser2_a>;
+					maxim,gmsl-dser-device = <&dser2>;
+                    is-prim-d4m;
+					ports {
+						#address-cells = <1>;
+						#size-cells = <0>;
+						port@0 {
+							reg = <0>;
+							d4m4_depth_out: endpoint {
+								port-index = <2>;
+								bus-width = <2>;
+								remote-endpoint = <&d4m4_depth_csi_in>;
+							};
+						};
+					};
 					mode0 {
-						mclk_khz = "37125";
-						num_lanes = "4";
-						tegra_sinterface = "serial_a";
-						vc_id = "0";
-						
-						discontinuous_clk = "no";
-						dpcm_enable = "false";
-						cil_settletime = "0";
-						
+						pixel_t = "grey_y16";
+						num_lanes = "2";
 						csi_pixel_bit_depth = "16";
-						mode_type = "yuv";
-						pixel_phase = "uyvy";
-
-						active_w = "1920";
-						active_h = "1536";
-						readout_orientation = "0";
-						line_length = "2200";
-						inherent_gain = "1";
-						
+						active_w = "1280";
+						active_h = "720";
+						tegra_sinterface = "serial_c";
+						mclk_khz = "24000";
 						pix_clk_hz = "74250000";
-						serdes_pix_clk_hz = "375000000";
-
-						gain_factor = "10";
-						min_gain_val = "1"; /* 0dB */
-						max_gain_val = "480"; /* 48dB */
-						step_gain_val = "3"; /* 0.3 */
-						default_gain = "1";
-						framerate_factor = "1000000";
-						min_framerate = "30000000"; /* 1.5 */
-						max_framerate = "30000000"; /* 30 */
-						step_framerate = "1";
-						default_framerate= "30000000";
-
-						exposure_factor = "1000000";
-						min_exp_time = "30"; /* us */
-						max_exp_time = "660000"; /* us */
-						step_exp_time = "1";
-						default_exp_time = "33334";/* us */
-						embedded_metadata_height = "0";
+						line_length = "1280"; /* 2200 */
+						embedded_metadata_height = "1";
+					};
+					gmsl-link {
+						src-csi-port = "b";
+						dst-csi-port = "a";
+						serdes-csi-link = "a";
+						csi-mode = "1x4";
+						st-vc = <0>;
+						vc-id = <0>;
+						num-lanes = <2>;
 					};
+				};
 
+				d4m4_ir: d4m4_ir@0c {
+					status = "ok";
+					def-addr = <0x10>;
+					def-d4m-addr = <0x1c>;
+					reg = <0x0c>;
+					compatible = "intel,d4xx";
+					vcc-supply = <&p3737_vdd_1v8_sys>;
+					cam-type = "Y8";
+					maxim,gmsl-ser-device = <&ser2_a>;
+					maxim,gmsl-dser-device = <&dser2>;
+                    prim-d4m = <&d4m4_depth>;
 					ports {
 						#address-cells = <1>;
 						#size-cells = <0>;
+
 						port@0 {
 							reg = <0>;
-							fzcam_fzcam_out0: endpoint {
-								vc-id = <0>;
-								port-index = <0>;
-								bus-width = <4>;
-								remote-endpoint = <&fzcam_csi_in0>;
-								};
+							d4m4_ir_out: endpoint {
+								port-index = <2>;
+								bus-width = <2>;
+								remote-endpoint = <&d4m4_ir_csi_in>;
 							};
 						};
+					};
+					/* mode0: Y8, mode1: depth D16 */
+					mode0 {
+						pixel_t = "grey_y16";
+						num_lanes = "2";
+						csi_pixel_bit_depth = "16";
+						active_w = "1280";
+						active_h = "720";
+						tegra_sinterface = "serial_c";
+						mclk_khz = "24000";
+						pix_clk_hz = "74250000";
+						line_length = "1280"; /* 2200 */
+						embedded_metadata_height = "0";
+					};
+					gmsl-link {
+						src-csi-port = "b";
+						dst-csi-port = "a";
+						serdes-csi-link = "a";
+						csi-mode = "1x4";
+						st-vc = <0>;
+						vc-id = <1>;
+						num-lanes = <2>;
+					};
 				};
 
-				fzcam_b@2b {
-					compatible = "nvidia,fzcam";
-					reg = <0x2b>;
-					
-					devnode = "video1";
-					sensor_model ="fzcam";
-
-					/* Define any required hw resources needed by driver */
-					/* ie. clocks, io pins, power sources */
-					clocks = <&bpmp_clks TEGRA234_CLK_EXTPERIPH1>,
-							<&bpmp_clks TEGRA234_CLK_EXTPERIPH1>;
-					clock-names = "extperiph1", "pllp_grtba";
-					mclk = "extperiph1";
-					channel = <0>;
+				d4m5_depth: d4m5_depth@2a {
+					status = "ok";
+					def-addr = <0x10>;
+					def-d4m-addr = <0x1d>;
+					reg = <0x2a>;
+					compatible = "intel,d4xx";
+					vcc-supply = <&p3737_vdd_1v8_sys>;
+					cam-type = "Depth";
+					maxim,gmsl-ser-device = <&ser2_b>;
+					maxim,gmsl-dser-device = <&dser2>;
+                    is-prim-d4m;
+					ports {
+						#address-cells = <1>;
+						#size-cells = <0>;
 
+						port@0 {
+							reg = <0>;
+							d4m5_depth_out: endpoint {
+								port-index = <2>;
+								bus-width = <2>;
+								remote-endpoint = <&d4m5_depth_csi_in>;
+							};
+						};
+					};
 					mode0 {
-						mclk_khz = "37125";
-						num_lanes = "4";
-						tegra_sinterface = "serial_a";
-						vc_id = "1";
-						
-						discontinuous_clk = "no";
-						dpcm_enable = "false";
-						cil_settletime = "0";
-						
+						pixel_t = "grey_y16";
+						num_lanes = "2";
 						csi_pixel_bit_depth = "16";
-						mode_type = "yuv";
-						pixel_phase = "uyvy";
-
-						active_w = "1920";
-						active_h = "1536";
-						readout_orientation = "0";
-						line_length = "2200";
-						inherent_gain = "1";
-						
+						active_w = "1280";
+						active_h = "720";
+						tegra_sinterface = "serial_c";
+						mclk_khz = "24000";
 						pix_clk_hz = "74250000";
-						serdes_pix_clk_hz = "375000000";
-
-						gain_factor = "10";
-						min_gain_val = "1"; /* 0dB */
-						max_gain_val = "480"; /* 48dB */
-						step_gain_val = "3"; /* 0.3 */
-						default_gain = "1";
-						framerate_factor = "1000000";
-						min_framerate = "30000000"; /* 1.5 */
-						max_framerate = "30000000"; /* 30 */
-						step_framerate = "1";
-						default_framerate= "30000000";
-						exposure_factor = "1000000";
-						min_exp_time = "30"; /* us */
-						max_exp_time = "660000"; /* us */
-						step_exp_time = "1";
-						default_exp_time = "33334";/* us */
-						embedded_metadata_height = "0";
+						line_length = "1280"; /* 2200 */
+						embedded_metadata_height = "1";
+					};
+					gmsl-link {
+						src-csi-port = "b";
+						dst-csi-port = "a";
+						serdes-csi-link = "b";
+						csi-mode = "1x4";
+						st-vc = <2>;
+						vc-id = <2>;
+						num-lanes = <2>;
 					};
+				};
 
+				d4m5_ir: d4m5_ir@2c {
+					status = "ok";
+					def-addr = <0x10>;
+					def-d4m-addr = <0x1d>;
+					reg = <0x2c>;
+					compatible = "intel,d4xx";
+					vcc-supply = <&p3737_vdd_1v8_sys>;
+					cam-type = "Y8";
+					maxim,gmsl-ser-device = <&ser2_b>;
+					maxim,gmsl-dser-device = <&dser2>;
+                    prim-d4m = <&d4m5_depth>;
 					ports {
 						#address-cells = <1>;
 						#size-cells = <0>;
+
 						port@0 {
 							reg = <0>;
-							fzcam_fzcam_out1: endpoint {
-								vc-id = <1>;
-								port-index = <0>;
-								bus-width = <4>;
-								remote-endpoint = <&fzcam_csi_in1>;
+							d4m5_ir_out: endpoint {
+								port-index = <2>;
+								bus-width = <2>;
+								remote-endpoint = <&d4m5_ir_csi_in>;
 							};
 						};
 					};
+					/* mode0: Y8, mode1: depth D16 */
+					mode0 {
+						pixel_t = "grey_y16";
+						num_lanes = "2";
+						csi_pixel_bit_depth = "16";
+						active_w = "1280";
+						active_h = "720";
+						tegra_sinterface = "serial_c";
+						mclk_khz = "24000";
+						pix_clk_hz = "74250000";
+						line_length = "1280"; /* 2200 */
+						embedded_metadata_height = "0";
+					};
+					gmsl-link {
+						src-csi-port = "b";
+						dst-csi-port = "a";
+						serdes-csi-link = "b";
+						csi-mode = "1x4";
+						st-vc = <2>;
+						vc-id = <3>;
+						num-lanes = <2>;
+					};
 				};
 			};
 		
-			i2c@3 { //MAX96712
+			i2c@3 { //MAX96712 - 4xISX031
 				reg = <3>;
 				#address-cells = <1>;
 				#size-cells = <0>;
@@ -897,11 +999,11 @@
 						#size-cells = <0>;
 						port@0 {
 							reg = <0>;
-							fzcam_fzcam_out2: endpoint {
+							isx031_0_out: endpoint {
 								vc-id = <0>;
 								port-index = <5>;
 								bus-width = <4>;
-								remote-endpoint = <&fzcam_csi_in2>;
+								remote-endpoint = <&isx031_0_csi_in>;
 							};
 						};
 					};
@@ -967,11 +1069,11 @@
 						#size-cells = <0>;
 						port@0 {
 							reg = <0>;
-							fzcam_fzcam_out3: endpoint {
+							isx031_1_out: endpoint {
 								vc-id = <1>;
 								port-index = <5>;
 								bus-width = <4>;
-								remote-endpoint = <&fzcam_csi_in3>;
+								remote-endpoint = <&isx031_1_csi_in>;
 							};
 						};
 					};
@@ -1038,11 +1140,11 @@
 						#size-cells = <0>;
 						port@0 {
 							reg = <0>;
-							fzcam_fzcam_out4: endpoint {
+							isx031_2_out: endpoint {
 								vc-id = <2>;
 								port-index = <5>;
 								bus-width = <4>;
-								remote-endpoint = <&fzcam_csi_in4>;
+								remote-endpoint = <&isx031_2_csi_in>;
 								};
 							};
 						};
@@ -1110,11 +1212,11 @@
 						#size-cells = <0>;
 						port@0 {
 							reg = <0>;
-							fzcam_fzcam_out5: endpoint {
+							isx031_3_out: endpoint {
 								vc-id = <3>;
 								port-index = <5>;
 								bus-width = <4>;
-								remote-endpoint = <&fzcam_csi_in5>;
+								remote-endpoint = <&isx031_3_csi_in>;
 							};
 						};
 					};
@@ -1129,7 +1231,7 @@
 			#size-cells = <0x0>;
 
 			num-channels = <14>;
-			//intel d4xx
+			//d4m0 Depth
 			channel@0 {
 				reg = <0x0>;
 				status = "ok";
@@ -1142,23 +1244,24 @@
 						reg = <0>;
 						status = "ok";
 
-						csi_in0: endpoint@0 {
+						d4m0_depth_csi_in: endpoint@0 {
 							status = "ok";
 							port-index = <4>;
 							bus-width = <2>;
-							remote-endpoint = <&d4m0_out>;
+							remote-endpoint = <&d4m0_depth_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
 						status = "ok";
-						csi_out0: endpoint@1 {
+						d4m0_depth_csi_out: endpoint@1 {
 							status = "ok";
-							remote-endpoint = <&vi_in0>;
+							remote-endpoint = <&d4m0_depth_vi_in>;
 						};
 					};
 				};
 			};
+			//d4m0 IR
 			channel@1 {
 				reg = <0x1>;
 				status = "ok";
@@ -1171,24 +1274,25 @@
 						reg = <0>;
 						status = "ok";
 
-						csi_in1: endpoint@2 {
+						d4m0_ir_csi_in: endpoint@2 {
 							status = "ok";
 							port-index = <4>;
 							bus-width = <2>;
-							remote-endpoint = <&d4m1_out>;
+							remote-endpoint = <&d4m0_ir_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
 						status = "ok";
-						csi_out1: endpoint@3 {
+						d4m0_ir_csi_out: endpoint@3 {
 							status = "ok";
-							remote-endpoint = <&vi_in1>;
+							remote-endpoint = <&d4m0_ir_vi_in>;
 						};
 					};
 
 				};
 			};
+			//d4m1 Depth
 			channel@2 {
 				reg = <0x2>;
 				status = "ok";
@@ -1201,23 +1305,24 @@
 						reg = <0>;
 						status = "ok";
 
-						csi_in2: endpoint@4 {
+						d4m1_depth_csi_in: endpoint@4 {
 							status = "ok";
 							port-index = <4>;
 							bus-width = <2>;
-							remote-endpoint = <&d4m2_out>;
+							remote-endpoint = <&d4m1_depth_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
 						status = "ok";
-						csi_out2: endpoint@5 {
+						d4m1_depth_csi_out: endpoint@5 {
 							status = "ok";
-							remote-endpoint = <&vi_in2>;
+							remote-endpoint = <&d4m1_depth_vi_in>;
 						};
 					};
 				};
 			};
+			//d4m1 IR
 			channel@3 {
 				reg = <0x3>;
 				status = "ok";
@@ -1230,23 +1335,24 @@
 						reg = <0>;
 						status = "ok";
 
-						csi_in3: endpoint@6 {
+						d4m1_ir_csi_in: endpoint@6 {
 							status = "ok";
 							port-index = <4>;
 							bus-width = <2>;
-							remote-endpoint = <&d4m3_out>;
+							remote-endpoint = <&d4m1_ir_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
 						status = "ok";
-						csi_out3: endpoint@7 {
+						d4m1_ir_csi_out: endpoint@7 {
 							status = "ok";
-							remote-endpoint = <&vi_in3>;
+							remote-endpoint = <&d4m1_ir_vi_in>;
 						};
 					};
 				};
 			};
+			//d4m2 Depth
 			channel@4 {
 				reg = <0x4>;
 				status = "ok";
@@ -1259,23 +1365,24 @@
 						reg = <0>;
 						status = "ok";
 
-						csi_in4: endpoint@8 {
+						d4m2_depth_csi_in: endpoint@8 {
 							status = "ok";
 							port-index = <2>;
 							bus-width = <2>;
-							remote-endpoint = <&d4m4_out>;
+							remote-endpoint = <&d4m2_depth_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
 						status = "ok";
-						csi_out4: endpoint@9 {
+						d4m2_depth_csi_out: endpoint@9 {
 							status = "ok";
-							remote-endpoint = <&vi_in4>;
+							remote-endpoint = <&d4m2_depth_vi_in>;
 						};
 					};
 				};
 			};
+			//d4m2 IR
 			channel@5 {
 				reg = <0x5>;
 				status = "ok";
@@ -1288,23 +1395,24 @@
 						reg = <0>;
 						status = "ok";
 
-						csi_in5: endpoint@10 {
+						d4m2_ir_csi_in: endpoint@10 {
 							status = "ok";
 							port-index = <2>;
 							bus-width = <2>;
-							remote-endpoint = <&d4m5_out>;
+							remote-endpoint = <&d4m2_ir_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
 						status = "ok";
-						csi_out5: endpoint@11 {
+						d4m2_ir_csi_out: endpoint@11 {
 							status = "ok";
-							remote-endpoint = <&vi_in5>;
+							remote-endpoint = <&d4m2_ir_vi_in>;
 						};
 					};
 				};
 			};
+			//d4m3 Depth
 			channel@6 {
 				reg = <0x6>;
 				status = "ok";
@@ -1317,23 +1425,24 @@
 						reg = <0>;
 						status = "ok";
 
-						csi_in6: endpoint@12 {
+						d4m3_depth_csi_in: endpoint@12 {
 							status = "ok";
 							port-index = <2>;
 							bus-width = <2>;
-							remote-endpoint = <&d4m6_out>;
+							remote-endpoint = <&d4m3_depth_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
 						status = "ok";
-						csi_out6: endpoint@13 {
+						d4m3_depth_csi_out: endpoint@13 {
 							status = "ok";
-							remote-endpoint = <&vi_in6>;
+							remote-endpoint = <&d4m3_depth_vi_in>;
 						};
 					};
 				};
 			};
+			//d4m3 IR
 			channel@7 {
 				reg = <0x7>;
 				status = "ok";
@@ -1346,147 +1455,229 @@
 						reg = <0>;
 						status = "ok";
 
-						csi_in7: endpoint@14 {
+						d4m3_ir_csi_in: endpoint@14 {
 							status = "ok";
 							port-index = <2>;
 							bus-width = <2>;
-							remote-endpoint = <&d4m7_out>;
+							remote-endpoint = <&d4m3_ir_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
 						status = "ok";
-						csi_out7: endpoint@15 {
+						d4m3_ir_csi_out: endpoint@15 {
 							status = "ok";
-							remote-endpoint = <&vi_in7>;
+							remote-endpoint = <&d4m3_ir_vi_in>;
 						};
 					};
 				};
 			};
 
-			//FZCAM
+			//d4m4 Depth
 			channel@8 {
-				reg = <8>;
+				reg = <0x8>;
+				status = "ok";
+
 				ports {
-					#address-cells = <1>;
-					#size-cells = <0>;
+					#address-cells = <0x1>;
+					#size-cells = <0x0>;
+
 					port@0 {
 						reg = <0>;
-						fzcam_csi_in0: endpoint@16 {
+						status = "ok";
+
+						d4m4_depth_csi_in: endpoint@16 {
+							status = "ok";
 							port-index = <0>;
-							bus-width = <4>;
-							remote-endpoint = <&fzcam_fzcam_out0>;
+							bus-width = <2>;
+							remote-endpoint = <&d4m4_depth_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
-						fzcam_csi_out0: endpoint@17 {
-							remote-endpoint = <&fzcam_vi_in0>;
+						status = "ok";
+						d4m4_depth_csi_out: endpoint@17 {
+							status = "ok";
+							remote-endpoint = <&d4m4_depth_vi_in>;
 						};
 					};
 				};
 			};
+			//d4m4 IR
 			channel@9 {
 				reg = <9>;
+				status = "ok";
+
 				ports {
-					#address-cells = <1>;
-					#size-cells = <0>;
+					#address-cells = <0x1>;
+					#size-cells = <0x0>;
+
 					port@0 {
 						reg = <0>;
-						fzcam_csi_in1: endpoint@18 {
+						status = "ok";
+
+						d4m4_ir_csi_in: endpoint@18 {
+							status = "ok";
 							port-index = <0>;
-							bus-width = <4>;
-							remote-endpoint = <&fzcam_fzcam_out1>;
+							bus-width = <2>;
+							remote-endpoint = <&d4m4_ir_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
-						fzcam_csi_out1: endpoint@19 {
-							remote-endpoint = <&fzcam_vi_in1>;
+						status = "ok";
+						d4m4_ir_csi_out: endpoint@19 {
+							status = "ok";
+							remote-endpoint = <&d4m4_ir_vi_in>;
 						};
 					};
+
 				};
 			};
+			//d4m5 Depth
 			channel@10 {
 				reg = <10>;
+				status = "ok";
+
+				ports {
+					#address-cells = <0x1>;
+					#size-cells = <0x0>;
+
+					port@0 {
+						reg = <0>;
+						status = "ok";
+
+						d4m5_depth_csi_in: endpoint@20 {
+							status = "ok";
+							port-index = <0>;
+							bus-width = <2>;
+							remote-endpoint = <&d4m5_depth_out>;
+						};
+					};
+					port@1 {
+						reg = <1>;
+						status = "ok";
+						d4m5_depth_csi_out: endpoint@21 {
+							status = "ok";
+							remote-endpoint = <&d4m5_depth_vi_in>;
+						};
+					};
+				};
+			};
+			//d4m5 IR
+			channel@11 {
+				reg = <11>;
+				status = "ok";
+
+				ports {
+					#address-cells = <0x1>;
+					#size-cells = <0x0>;
+
+					port@0 {
+						reg = <0>;
+						status = "ok";
+
+						d4m5_ir_csi_in: endpoint@22 {
+							status = "ok";
+							port-index = <0>;
+							bus-width = <2>;
+							remote-endpoint = <&d4m5_ir_out>;
+						};
+					};
+					port@1 {
+						reg = <1>;
+						status = "ok";
+						d4m5_ir_csi_out: endpoint@23 {
+							status = "ok";
+							remote-endpoint = <&d4m5_ir_vi_in>;
+						};
+					};
+				};
+			};
+			// ISX031 #0
+			channel@12 {
+				reg = <12>;
 				ports {
 					#address-cells = <1>;
 					#size-cells = <0>;
 					port@0 {
 						reg = <0>;
-						fzcam_csi_in2: endpoint@20 {
+						isx031_0_csi_in: endpoint@24 {
 							port-index = <6>;
 							bus-width = <4>;
-							remote-endpoint = <&fzcam_fzcam_out2>;
+							remote-endpoint = <&isx031_0_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
-						fzcam_csi_out2: endpoint@21 {
-							remote-endpoint = <&fzcam_vi_in2>;
+						isx031_0_csi_out: endpoint@25 {
+							remote-endpoint = <&isx031_0_vi_in>;
 						};
 					};
 				};
 			};
-			channel@11 {
-				reg = <11>;
+			// ISX031 #1
+			channel@13 {
+				reg = <13>;
 				ports {
 					#address-cells = <1>;
 					#size-cells = <0>;
 					port@0 {
 						reg = <0>;
-						fzcam_csi_in3: endpoint@22 {
+						isx031_1_csi_in: endpoint@26 {
 							port-index = <6>;
 							bus-width = <4>;
-							remote-endpoint = <&fzcam_fzcam_out3>;
+							remote-endpoint = <&isx031_1_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
-						fzcam_csi_out3: endpoint@23 {
-							remote-endpoint = <&fzcam_vi_in3>;
+						isx031_1_csi_out: endpoint@27 {
+							remote-endpoint = <&isx031_1_vi_in>;
 						};
 					};
 				};
-			};			
-			channel@12 {
-				reg = <12>;
+			};
+			// ISX031 #2
+			channel@14 {
+				reg = <14>;
 				ports {
 					#address-cells = <1>;
 					#size-cells = <0>;
 					port@0 {
 						reg = <0>;
-						fzcam_csi_in4: endpoint@24 {
+						isx031_2_csi_in: endpoint@28 {
 							port-index = <6>;
 							bus-width = <4>;
-							remote-endpoint = <&fzcam_fzcam_out4>;
+							remote-endpoint = <&isx031_2_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
-						fzcam_csi_out4: endpoint@25{
-							remote-endpoint = <&fzcam_vi_in4>;
+						isx031_2_csi_out: endpoint@29 {
+							remote-endpoint = <&isx031_2_vi_in>;
 						};
 					};
 				};
 			};
-			channel@13 {
-				reg = <13>;
+			// ISX031 #3
+			channel@15 {
+				reg = <15>;
 				ports {
 					#address-cells = <1>;
 					#size-cells = <0>;
 					port@0 {
 						reg = <0>;
-						fzcam_csi_in5: endpoint@26 {
+						isx031_3_csi_in: endpoint@30 {
 							port-index = <6>;
 							bus-width = <4>;
-							remote-endpoint = <&fzcam_fzcam_out5>;
+							remote-endpoint = <&isx031_3_out>;
 						};
 					};
 					port@1 {
 						reg = <1>;
-						fzcam_csi_out5: endpoint@27 {
-							remote-endpoint = <&fzcam_vi_in5>;
+						isx031_3_csi_out: endpoint@31 {
+							remote-endpoint = <&isx031_3_vi_in>;
 						};
 					};
 				};
